r2i/onnxrt/engine.cc

```diff
@@ -0,0 +1,281 @@
+    return nullptr;
+  }
+
+  return prediction;
+}
+
+RuntimeError Engine::ValidateInputTensorShape (int channels, int height,
+    int width, std::vector<int64_t> input_dims) {
+  RuntimeError error;
+
+  // We only support 1 batch
+  if (1 != input_dims.at(0)) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "We only support a batch of 1 image(s) in our frames");
+    return error;
+  }
+
+  // Check that channels match
+  if (channels != input_dims.at(1)) {
+    std::string error_msg;
+    error_msg = "Channels per image:" + std::to_string(channels) +
+                ", needs to be equal to model input channels:" +
+                std::to_string(input_dims.at(1));
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER, error_msg);
+    return error;
+  }
+
+  // Check that heights match
+  if (height != input_dims.at(2)) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Unsupported image height");
+    return error;
+  }
+
+  // Check that widths match
+  if (width != input_dims.at(3)) {
+    error.Set (RuntimeError::Code::INVALID_FRAMEWORK_PARAMETER,
+               "Unsupported image width");
+    return error;
+  }
+
+  return error;
+}
+
+RuntimeError Engine::ScoreModel (std::shared_ptr<Ort::Session> session,
+                                 std::shared_ptr<Frame> frame,
+                                 size_t input_image_size,
+                                 std::vector<int64_t> input_node_dims,
+                                 std::shared_ptr<Prediction> prediction) {
+  RuntimeError error;
+  float *result;
+  int64_t output_size;
+  size_t num_input_nodes = session->GetInputCount();
+  size_t num_output_nodes = session->GetOutputCount();
+  std::vector<int64_t> output_node_dims;
+  Ort::Value input_tensor{nullptr};
+  Ort::AllocatorWithDefaultOptions input_allocator;
+  Ort::AllocatorWithDefaultOptions output_allocator;
+
+  // Warning: We support batches of size 1 and models with
+  // 1 output only.
+  try {
+    output_node_dims = session->GetOutputTypeInfo(
+                         0).GetTensorTypeAndShapeInfo().GetShape();
+    output_size = session->GetOutputTypeInfo(
+                    0).GetTensorTypeAndShapeInfo().GetElementCount();
+  }
+
+  catch (std::exception &excep) {
+    error.Set(RuntimeError::Code::FRAMEWORK_ERROR, excep.what());
+    return error;
+  }
+
+  num_input_nodes = session->GetInputCount();
+  num_output_nodes = session->GetOutputCount();
+
+  std::vector<const char *> input_node_names(num_input_nodes);
+  std::vector<const char *> output_node_names(num_output_nodes);
+
+  // Warning: only 1 input and output supported
+  char *input_name = session->GetInputName(0, input_allocator);
+  input_node_names[0] = input_name;
+  char *output_name = session->GetOutputName(0, output_allocator);
+  output_node_names[0] = output_name;
+
+  try {
+    Ort::MemoryInfo memory_info = Ort::MemoryInfo::CreateCpu(OrtArenaAllocator,
```

Comment:
Do we need to do this with every iteration? Can't we save it internally?

Reply:
No, we don't need to do this in every iteration. I'll factor this out.